src/datadog/remote_config/remote_config.cpp

```diff
@@ -45,68 +29,74 @@ constexpr std::array<uint8_t, sizeof(uint64_t)> capabilities_byte_array(
-    config_update.tags = tags_it->get<std::vector<StringView>>();
-  }
+  assert(match.ready());
+  assert(match.size() == 3);
 
-  if (auto tracing_enabled_it = j.find("tracing_enabled");
-      tracing_enabled_it != j.cend() && tracing_enabled_it->is_boolean()) {
-    config_update.report_traces = tracing_enabled_it->get<bool>();
-  }
-
-  if (auto tracing_sampling_rules_it = j.find("tracing_sampling_rules");
-      tracing_sampling_rules_it != j.cend() &&
-      tracing_sampling_rules_it->is_array()) {
-    config_update.trace_sampling_rules = &(*tracing_sampling_rules_it);
-  }
+  StringView product_sv(config_path.data() + match.position(2),
+                        std::size_t(match.length(2)));
 
-  return config_update;
+  return parse_product(product_sv);
 }
 
 }  // namespace
 
-RemoteConfigurationManager::RemoteConfigurationManager(
-    const TracerSignature& tracer_signature,
-    const std::shared_ptr<ConfigManager>& config_manager)
+Manager::Manager(const TracerSignature& tracer_signature,
+                 const std::vector<std::shared_ptr<Listener>>& listeners)
     : tracer_signature_(tracer_signature),
-      config_manager_(config_manager),
+      listeners_(listeners),
       client_id_(uuid()) {
-  assert(config_manager_);
+  Capabilities capabilities = 0;
+
+  for (const auto& listener : listeners_) {
+    visit_products(
+        listener->get_products(), [this, &listener](product::Flag product) {
+          products_.emplace(to_string_view(product));
+          listeners_per_product_[product].emplace_back(listener.get());
+        });
+    capabilities |= listener->get_capabilities();
+  }
+
+  // The ".client.capabilities" field of the remote config request payload
+  // describes which parts of the library's configuration are supported for
+  // remote configuration.
+  //
+  // It's a bitset, 64 bits wide, where each bit indicates whether the library
+  // supports a particular feature for remote configuration.
+  //
+  // The bitset is encoded in the request as a JSON array of 8 integers, where
+  // each integer is one byte from the 64 bits. The bytes are in big-endian
+  // order within the array.
+  capabilities_ = capabilities_byte_array(capabilities);
 }
 
-bool RemoteConfigurationManager::is_new_config(
-    StringView config_path, const nlohmann::json& config_meta) {
+bool Manager::is_new_config(StringView config_path,
+                            const nlohmann::json& config_meta) {
   auto it = applied_config_.find(std::string{config_path});
   if (it == applied_config_.cend()) return true;
 
   return it->second.hash !=
          config_meta.at("/hashes/sha256"_json_pointer).get<StringView>();
 }
 
-nlohmann::json RemoteConfigurationManager::make_request_payload() {
+nlohmann::json Manager::make_request_payload() {
   // clang-format off
   auto j = nlohmann::json{
     {"client", {
       {"id", client_id_},
-      {"products", nlohmann::json::array({k_apm_product})},
+      {"products", products_},
```

Comment:
You don't seem to be using `cached_target_files`, which results in the same configuration being sent over and over again.

Reply:
I deferred this until I am able to get an answer from RC people. My observation is configurations are not sent over and over again.

Reply:
I think the RFC is quite clear here:

> Used to avoid sending the same config files over and over on the wire. Note that the agent does not make any effort trying to track this for the tracer and will resend every file in every update if this array is empty.

I don't think observations are relevant, the remote configuration client must meet the specification regardless of the idiosyncrasies of the agent.

Reply:
addresed in d547482a2d932db2a50043836caf8353aa4ae12b